### ISSUE TITLE
Bitnami Redis Sentinel & Sidekiq on Kubernetes

### ISSUE THREAD

**Issue by h0jeZvgoxFepBQ2C** [mperham/sidekiq#5194]
I know this partly off topic, but I still hope to find someone who maybe knows how to proceed:

So right now we are trying to get a [Bitnami Redis Sentinel](https://artifacthub.io/packages/helm/bitnami/redis) cluster working, together with our Rails app + Sidekiq.

We tried different things, but it's not really clear to us, how we should specify the sentinels for Sidekiq (crutial part is here, that the sentinel nodes are READ ONLY, so we cannot use them for sidekiq, since job states get written).

Since on Kubernetes there are only 2 services available: "redis" and "redis-headless" (not sure how they differ?) - how can I specify the sentinels like this:

```ruby
Sidekiq.configure_server do |config|
  config.redis = {
    url: "redis",
    sentinels: [
      { host: "?", port: 26379 } # why do we have to specify it here seperately, since we should be able to get a unified answer via a service, or?
      { host: "?", port: 26379 }  
      { host: "?", port: 26379 } 
    }
  }
end
```

Would be nice if someone can shed some light on this. As far as I understood, the bitnami redis sentiel only returns the IP of the master and the application has to handle the corresponding writes to this master then (https://github.com/bitnami/charts/tree/master/bitnami/redis#master-replicas-with-sentinel) - but I really don't understand on how to do this with sidekiq?

**Comment by h0jeZvgoxFepBQ2C** [comment]
Maybe we should add this information somewhere to the Readme or Wiki?

**Comment by mperham** [comment]
This is adequate documentation as folks can search and find this. Most users aren't using Sentinel; the wiki is not meant to document every last edge case.

**Comment by grjones** [comment]
@h0jeZvgoxFepBQ2C I don't think this feels right. You'd ultimately be setting explicit IP address in your config. What if these change because your redis pods are recreated?  Why not use the regular (non-headless) service that will connect you to an available sentinel?

**Comment by itay-grudev** [comment]
@grjones I get your point and I thought about it and I think you are partially correct.

The non-headless service will return a POD address, which may fail. Meaning you would have a connection only to a single failed pod, albeit a current one (so a pod that failed recently and is not yet restarted).

The headless service will return a pods at the time of calling, but it will return a bunch of them - so you will have redundancy.

So the solution that forms in my mind looks like this:

```ruby
hostnames = [ ENV['REDIS_SERVICE'] ]
hostnames += Resolv.getaddresses( ENV['REDIS_SERVICE_HEADLESS'] )

sentinels = hostnames.map do |hostname|
  {
    host: hostname,
    port: 26379,
    password: ENV['REDIS_PASSWORD'],
  }
end

$redis = Redis.new(
  url: 'redis://mymaster',
  password: ENV['REDIS_PASSWORD'],
  name: 'mymaster',
  sentinels: sentinels,
  role: :master,
)

$redisRead = Redis.new(
  url: 'redis://mymaster',
  password: ENV['REDIS_PASSWORD'],
  name: 'mymaster',
  sentinels: sentinels,
  role: :slave,
)
```

This way you use multiple sentinel connections, referenced by their IPs and once using the non-headless service.

**Comment by itay-grudev** [comment]
In fact this would probably be the best approach:

```ruby
hostnames = [ ENV['REDIS_SERVICE'], ENV['REDIS_SERVICE_HEADLESS'] ]
hostnames += Resolv.getaddresses( ENV['REDIS_SERVICE_HEADLESS'] )
```

**Comment by alzabo** [comment]
The Bitnami chart uses a Kubernetes StatefulSet to manage the Pods running the Sentinel service. StatefulSet Pods have a predictable DNS name that may be used instead of resolving the Pod IPs from the headless service DNS name.

The format of the pod DNS records is `$(statefulset name)-$(ordinal)`.

The domain uses the format `$(service name).$(namespace).svc.cluster.local`. The service name is specified in the StatefulSet's `.spec.serviceName`.

Official docs for this StatefulSet feature: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id.

Redis/Sentinel Pod IPs will change when they are rescheduled, which may occur for any number of reasons. Consider Kubernetes Node upgrades, which cause all pods to be rescheduled. If your client and server pods are started in the wrong order the clients will not be able to connect.
